compiler/src/dotty/tools/dotc/reporting/WConf.scala

```diff
@@ -0,0 +1,123 @@
+    case (filters, action) if filters.forall(_.matches(message)) => action
+  }.getOrElse(Action.Warning)
+
+object WConf:
+  import Action._
+  import MessageFilter._
+
+  private type Conf = (List[MessageFilter], Action)
+
+  def parseAction(s: String): Either[List[String], Action] = s match {
+    case "error" | "e"            => Right(Error)
+    case "warning" | "w"          => Right(Warning)
+    case "verbose" | "v"          => Right(Verbose)
+    case "info" | "i"             => Right(Info)
+    case "silent" | "s"           => Right(Silent)
+    case _                        => Left(List(s"unknown action: `$s`"))
+  }
+
+  private def regex(s: String) =
+    try Right(s.r)
+    catch { case e: PatternSyntaxException => Left(s"invalid pattern `$s`: ${e.getMessage}") }
+
+  @sharable val Splitter = raw"([^=]+)=(.+)".r
+  @sharable val ErrorId = raw"E?(\d+)".r
+
+  def parseFilter(s: String): Either[String, MessageFilter] = s match
+    case "any" => Right(Any)
+    case Splitter(filter, conf) => filter match
+      case "msg" => regex(conf).map(MessagePattern.apply)
+      case "id" => conf match
+        case ErrorId(num) =>
+          val n = num.toInt + 2
+          if n < ErrorMessageID.values.length then
+            Right(MessageID(ErrorMessageID.fromOrdinal(n)))
+          else
+            Left(s"unknonw error message id: E$n")
+        case _ =>
+          Left(s"invalid error message id: $conf")
+      case "name" =>
+        try Right(MessageID(ErrorMessageID.valueOf(conf + "ID")))
+        catch case _: IllegalArgumentException => Left(s"unknown error message name: $conf")
+
+      case "cat" => conf match
+        case "deprecation" => Right(Deprecated)
+        case "feature"     => Right(Feature)
+        case "unchecked"   => Right(Unchecked)
+        case _             => Left(s"unknown category: $conf")
+      case _ => Left(s"unknown filter: $filter")
+    case _ => Left(s"unknown filter: $s")
+
+  def parsed(using Context): WConf =
+    val setting = ctx.settings.Wconf.value
+    def cached = ctx.base.wConfCache
+    if cached == null || cached._1 != setting then
+      val conf = fromSettings(setting)
+      ctx.base.wConfCache = (setting, conf.getOrElse(WConf(Nil)))
+      conf.swap.foreach(msgs =>
+        val multiHelp =
+          if setting.sizeIs > 1 then
+            """
+              |Note: for multiple filters, use `-Wconf:filter1:action1,filter2:action2`
+              |      or alternatively          `-Wconf:filter1:action1 -Wconf:filter2:action2`""".stripMargin
+          else ""
+        report.warning(s"Failed to parse `-Wconf` configuration: ${ctx.settings.Wconf.value.mkString(",")}\n${msgs.mkString("\n")}$multiHelp"))
+    cached._2
+
+  def fromSettings(settings: List[String]): Either[List[String], WConf] =
+    if (settings.isEmpty) Right(WConf(Nil))
+    else {
+      val parsedConfs: List[Either[List[String], (List[MessageFilter], Action)]] = settings.map(conf => {
```

Comment:
Actually I would recommend writing `settings.map(conf => ...` without using `{` at all, it is not necessary since `=>` can introduce an indentation region.